### PR TITLE
[#11825] Warn user before deleting last course

### LINKED ISSUE #11825: Warn user before deleting last course in institute
- **Environment**: live server at V8.15.0

**Description of feature/enhancement**

- Warn user if s/he is about to delete his/her last course within an institute, as it will cause him/her to lose status as instructor in the said institute.
- (Optional) warn user if s/he is about to delete his/her last course or all of his/her courses, as s/he will lose access to the system as an instructor.

**Justification**

It is a design that an instructor can only add new courses in an institute whereby s/he is already a course co-owner. However, users are not expected to be aware of this detail, and they will expect some form of warning if they are about to lose some form of access.

### PR DESCRIPTION
**Pull request by inifnite**
Fixes #11825

**PR Checklist**

**Outline of Solution**
An instructor can only lose access when the courses are deleted permanently. So, in the "instructor-courses-page.component.ts" , I checked if the course being deleted permanently is the last course in that institute or overall and showed the warning accordingly.

### CODE REVIEW

**Review by zhaojj2209 — CHANGES_REQUESTED** [review]
The current implementation causes both a warning toast and a warning modal to pop up when the user tries to delete the last course permanently; I believe a better solution would be to add the warning message in the existing modal, with more emphasis placed (e.g. bolding the text, changing text colour to red) if needed.

image

**Comment by inifnite** [comment]
Okay. I will make the said changes.

**Comment by inifnite** [comment]
@zhaojj2209 @damithc is this fine?

https://user-images.githubusercontent.com/45487930/175760010-b5d670e1-22cb-43d4-a776-760135a8750f.mov

**Comment by damithc** [comment]
> @zhaojj2209 @damithc is this fine?

I'm OK with this UX (others welcome to suggest alternatives). The second confirmation dialog need to emphasize the issue at hand, which is losing access e.g., the title of the dialog box should be revised accordingly.

**Comment by inifnite** [comment]
Below is the changed dialog box, have a look at it.
Also should we add losing access warning for deleteAll courses from Recycle bin  operation  as well?? 
Screenshot 2022-06-25 at 4 34 20 PM

**Comment by damithc** [comment]
How about something like this?

> **This action will cause you to lose access to TEAMMATES!**
>--------
>If you delete _all_ courses of institute 'ABC', you will lose instructor access to TEAMMATES under the institution 'ABC'. To retain access, ensure you keep at least one course for each institution you are an instructor of. Are you sure you want to delete the course XYZ in institution 'ABC'?

Yes, the warning needs to be shown for 'delete all' as well, if applicable.

**Comment by inifnite** [comment]
This is your last course registered on TEAMMATES for which you have instructor access. Deleting this course will remove your instructor access to TEAMMATES.
Are you sure you want to delete the course a11.a.a-demo in institution a?

Would this work in case of deleting last course of all courses??

**Comment by inifnite** [comment]
Is this okay??

https://user-images.githubusercontent.com/45487930/175782765-f0522052-d6a6-45b5-ba7f-79f850225cc1.mov

**Comment by damithc** [comment]
@inifnite I'm OK with this. Perhaps use red text for `loose access` (or highlight using yellow background)? I would like the second dialog to appear visibly different from the first one in an eye-catching way such that the user immediately realizes this is different and possibly more serious than the previous one.

**Comment by inifnite** [comment]
for making second dialogue different from first, first has to be changed to warning and second to be remained as danger. Also, I tried changing color for `loose access` but it did not work. I just put sort of yellow background. have a look and let me know if this works??
Screenshot 2022-06-28 at 9 43 53 AM
Screenshot 2022-06-28 at 9 44 04 AM

**Comment by damithc** [comment]
> for making second dialogue different from first, first has to be changed to warning and second to be remained as danger. Also, I tried changing color for `loose access` but it did not work. I just put sort of yellow background. have a look and let me know if this works??
> Screenshot 2022-06-28 at 9 43 53 AM


Best not to change the style of this one, as its style should be consistent with other delete confirmations across TEAMMATES.

> Screenshot 2022-06-28 at 9 44 04 AM

Also highlight `remove your instructor access` for good measure?

**Comment by inifnite** [comment]
would this do??

https://user-images.githubusercontent.com/45487930/176353438-6f9b7b00-0bcf-4609-a57d-8eb98be23e11.mov

**Comment by inifnite** [comment]
@zhaojj2209 Please have a look.

**Review by zhaojj2209 — CHANGES_REQUESTED** [review]
Code works, but code quality needs work. Some general comments:
- Ensure consistent code indentation
- Declaring constants is preferred over declaring variables and then reassigning them

**Comment by inifnite** [comment]
@zhaojj2209 using else while having return in if block, gives lint error

File: src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts [hunk 1]
```diff
@@ -499,11 +534,26 @@ export class InstructorCoursesPageComponent implements OnInit {
...
+
+        return this.simpleModalService.openConfirmationModal(
+          'This action will cause you to <mark><strong>lose access</strong></mark> to TEAMMATES!',
+          SimpleModalType.DANGER, modalContentCnf).result;
+      }
+      return Promise.resolve();
+    }).then(() => {
       const deleteRequests: Observable<MessageOutput>[] = [];
-      this.softDeletedCourses.forEach((courseToDelete: CourseModel) => {
+        this.softDeletedCourses.forEach((courseToDelete: CourseModel) => {
```
[the remark this answers is not in the document]
Comment (zhaojj2209) [reply src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:546]: Almost there! This suggested change was missed out.

**Comment by inifnite** [comment]
Sorry for the delay

**Review by NicolasCwy — APPROVED** [review]
Changes look good to me! Nice work

**Review by zhaojj2209 — APPROVED** [review]
LGTM, thanks for the changes!

**Pull request merged**